File: backend/app/services/geo.py
"""Geo helpers.

Hotspot aggregation uses a fixed-size lat/lon grid rather than a clustering
library: cells are stable across queries (so a hotspot keeps its identity as
cases accumulate), cheap to index in SQL, and trivially explainable to an
agriculture officer -- "this 5km square has 14 confirmed late blight cases".
"""
from __future__ import annotations

import math

# ~0.05 deg latitude is ~5.5 km. Coarse enough to form clusters from sparse
# field reports, fine enough to point an extension officer at a village group.
DEFAULT_CELL_DEG = 0.05
EARTH_RADIUS_KM = 6371.0


def geo_cell(lat: float | None, lon: float | None, size_deg: float = DEFAULT_CELL_DEG) -> str | None:
    if lat is None or lon is None:
        return None
    lat_i = math.floor(lat / size_deg)
    lon_i = math.floor(lon / size_deg)
    return f"{size_deg:g}:{lat_i}:{lon_i}"


def cell_center(cell: str) -> tuple[float, float]:
    size_s, lat_s, lon_s = cell.split(":")
    size = float(size_s)
    return (int(lat_s) + 0.5) * size, (int(lon_s) + 0.5) * size


def haversine_km(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    p1, p2 = math.radians(lat1), math.radians(lat2)
    dp = p2 - p1
    dl = math.radians(lon2 - lon1)
    a = math.sin(dp / 2) ** 2 + math.cos(p1) * math.cos(p2) * math.sin(dl / 2) ** 2
    return 2 * EARTH_RADIUS_KM * math.asin(math.sqrt(a))

File: mobileapp/app/lib/domain/geo.dart
/// Port of backend/app/services/geo.py.
///
/// Validated against mobileapp/fixtures/geo.json.
library;

import 'dart:math' as math;

/// ~0.05 deg latitude is ~5.5 km - coarse enough to cluster sparse field
/// reports, fine enough to point an officer at a village group.
const double kDefaultCellDeg = 0.05;
const double kEarthRadiusKm = 6371.0;

/// Grid-cell id for a coordinate, or null if either axis is missing.
///
/// Uses floor, not truncation. `(-1.02 / 0.05).toInt()` is -20 in Dart but
/// Python's `math.floor` gives -21; every coordinate in Maharashtra is
/// positive so a truncating port passes every manual test and silently
/// mis-buckets the moment the app is used south of the equator or west of
/// Greenwich.
String? geoCell(double? lat, double? lon, {double sizeDeg = kDefaultCellDeg}) {
  if (lat == null || lon == null) return null;
  final latI = (lat / sizeDeg).floor();
  final lonI = (lon / sizeDeg).floor();
  return '${_g(sizeDeg)}:$latI:$lonI';
}

/// Centre of a cell, as (lat, lon).
({double lat, double lon}) cellCenter(String cell) {
  final parts = cell.split(':');
  final size = double.parse(parts[0]);
  final latI = int.parse(parts[1]);
  final lonI = int.parse(parts[2]);
  return (lat: (latI + 0.5) * size, lon: (lonI + 0.5) * size);
}

/// Great-circle distance in kilometres.
double haversineKm(double lat1, double lon1, double lat2, double lon2) {
  final p1 = _rad(lat1);
  final p2 = _rad(lat2);
  final dp = p2 - p1;
  final dl = _rad(lon2 - lon1);
  final a = math.sin(dp / 2) * math.sin(dp / 2) +
      math.cos(p1) * math.cos(p2) * math.sin(dl / 2) * math.sin(dl / 2);
  return 2 * kEarthRadiusKm * math.asin(math.sqrt(a));
}

double _rad(double deg) => deg * math.pi / 180.0;

/// Mirrors Python's `'%g' % size` formatting of the cell-size prefix, so cell
/// ids generated here are byte-identical to ones generated server-side
/// (0.05 -> "0.05", 0.25 -> "0.25", 1.0 -> "1").
String _g(double v) {
  if (v == v.roundToDouble() && v.abs() < 1e6) {
    return v.toInt().toString();
  }
  var s = v.toString();
  if (s.contains('.')) {
    s = s.replaceFirst(RegExp(r'0+$'), '');
    s = s.replaceFirst(RegExp(r'\.$'), '');
  }
  return s;
}
